sklearn/decomposition/online_lda.py

```diff
@@ -0,0 +1,668 @@
+        if self.n_jobs == -1:
+            n_jobs = cpu_count()
+        else:
+            n_jobs = self.n_jobs
+
+        results = Parallel(n_jobs=n_jobs, verbose=self.verbose)(
+            delayed(_update_doc_distribution)
+            (X[idx_slice, :], self.exp_dirichlet_component_, self.doc_topic_prior_,
+             self.max_doc_update_iter, self.mean_change_tol, cal_diff)
+            for idx_slice in gen_even_slices(X.shape[0], n_jobs))
+
+        # merge result
+        doc_topics, comp_diffs = zip(*results)
+        doc_topic_distr = np.vstack(doc_topics)
+
+        if cal_diff:
+            # This step finishes computing the sufficient statistics for the M step
+            component_diff = np.zeros(self.components_.shape)
+            for comp_diff in comp_diffs:
+                component_diff += comp_diff
+            component_diff *= self.exp_dirichlet_component_
+        else:
+            component_diff = None
+
+        return (doc_topic_distr, component_diff)
+
+    def _em_step(self, X, total_samples, batch_update):
+        """
+        EM update for 1 iteration
+        update `_component` by bath VB or online VB
+
+        parameters
+        ----------
+        X : sparse matrix, shape=(n_samples, n_features)
+            Document word matrix.
+
+        total_samples : integer
+            Total umber of document. It is only used when
+            batch_update is `False`.
+
+        batch_update : boolean
+            parameter that control updating method.
+            `True` for batch learning, `False` for online learning
+
+        Returns
+        -------
+        doc_topic_distr : array, shape=(n_samples, n_topics)
+            Unnormailzed document topic distribution.
+        """
+
+        # E-step
+        doc_topic_distr, component_diff = self._e_step(X, cal_diff=True)
+
+        # M-step
+        if batch_update:
+            self.components_ = self.topic_word_prior_ + component_diff
+        else:
+            # online update
+            # In the literature, the weight is `rho`
+            weight = np.power(self.learning_offset + self.n_iter_, -self.learning_decay)
+            doc_ratio = float(total_samples) / X.shape[0]
+            self.components_ *= (1 - weight)
+            self.components_ += (weight *
+                                 (self.topic_word_prior_ + doc_ratio * component_diff))
+
+        # update `component_` related variables
+        self.dirichlet_component_ = _dirichlet_expectation(self.components_)
+        self.exp_dirichlet_component_ = np.exp(self.dirichlet_component_)
+        self.n_iter_ += 1
+        return doc_topic_distr
+
+    def _to_csr(self, X, whom):
+        """
+        check & convert X to csr format, and make sure no negative value in X.
+
+        parameters
+        ----------
+        X :  array-like
+        
+        """
+        X = check_array(X, accept_sparse='csr')
+        check_non_negative(X, whom)
+        if not sp.issparse(X):
```

Comment:
Ah, so my comment about dense matrices was not entirely correct.... The sparse format is only used in a few places, right? Have you tried running on dense data?